cloudinit/cmd/daemon.py

```diff
@@ -0,0 +1,109 @@
+# This file is part of cloud-init. See LICENSE file for license information.
+
+""" Run cloud-init in daemon mode. """
+
+import argparse
+import contextlib
+import json
+import os
+import socket
+import sys
+
+from cloudinit.cmd.main import main as cimain
+
+NAME = 'daemon'
+CI_SOCKET = "/run/cloud-init/unix.socket"
+
+
+@contextlib.contextmanager
+def DomainServer(addr):
+    sock = None
+    try:
+        basedir = os.path.dirname(addr)
+        if not os.path.exists(basedir):
+            os.makedirs(basedir)
+        if os.path.exists(addr):
+            os.unlink(addr)
+        sock = socket.socket(socket.AF_UNIX, socket.SOCK_STREAM)
+        sock.bind(addr)
+        sock.listen(10)
+        yield sock
+    finally:
+        if sock:
+            sock.close()
+        if os.path.exists(addr):
+            os.unlink(addr)
+
+
+def log(stream, msg):
+    stream.write(msg + '\n')
+    stream.flush()
+
+
+def info(msg):
+    log(sys.stdout, "INFO: " + str(msg))
+
+
+def error(msg):
+    log(sys.stderr, "ERROR: " + str(msg))
+
+
+def handle_args(name, args):
+    with DomainServer(CI_SOCKET) as sock:
+        expected = set(['local', 'net', 'modules', 'final'])
+        completed = set()
```

Comment:
maybe completed needs to be sourced as well from /run/cloud-init/status.json if present and finished != None for that stage. otherwise the daemon, if restarted due to failure, would never close() as completed would never contain whichever stages already completed before daemon restart

Reply:
Daemon won't restart, however; it should check the same stuff that cloud-init status checks; I may even just invoke/use cloudinit/cmd/status.py:_get_status_details.